workspaces/cli-shared/src/captures/avro/file-system/capture-saver.ts

```diff
@@ -44,12 +52,14 @@ export class CaptureSaver implements ICaptureSaver {
         batchId,
       };
       try {
-        await this.onBatch(
-          groupingIdentifiers,
-          batchId,
-          items,
-          outputDirectory
-        );
+        await this.throttler.schedule(async () => {
```

Comment:
I think you want to put const promise = this.onBatch(...) , and then schedule(() => promise) and then await promise. that way, the flushing done in onBatch proceeds immediately, and the recordkeeping of scheduled work happens on its own separately